kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt

```diff
@@ -16,52 +16,198 @@
+import org.jetbrains.kotlin.descriptors.ClassKind.INTERFACE
+import org.jetbrains.kotlin.descriptors.ClassKind.OBJECT
+import org.jetbrains.kotlin.descriptors.DeclarationDescriptor
+import org.jetbrains.kotlin.descriptors.DeclarationDescriptorWithVisibility
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.descriptors.ValueDescriptor
+import org.jetbrains.kotlin.idea.kdoc.findKDoc
+import org.jetbrains.kotlin.load.java.lazy.descriptors.LazyJavaPackageFragment
+import org.jetbrains.kotlin.psi.KtDeclaration
 import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.resolve.BindingContext.REFERENCE_TARGET
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.resolve.descriptorUtil.isExtension
+import org.jetbrains.kotlin.resolve.descriptorUtil.isExtensionProperty
+import org.jetbrains.kotlin.resolve.lazy.descriptors.LazyClassDescriptor
 import org.jetbrains.kotlin.ui.editors.KotlinEditor
+import org.jetbrains.kotlin.ui.editors.navigation.findDeclarationInParsedFile
+import org.jetbrains.kotlin.ui.editors.navigation.getKtFileFromElement
+import org.jetbrains.kotlin.ui.editors.quickassist.resolveToDescriptor
+import java.net.URL
 
-class KotlinJavadocTextHover : KotlinEditorTextHover<BrowserInformationControlInput> {
+class KotlinJavadocTextHover : KotlinEditorTextHover<BrowserInformationControlInput>, ITextHoverExtension {
+
+    class KotlinInformationControlCreator : IInformationControlCreator {
+
+        override fun createInformationControl(parent: Shell?): IInformationControl =
+            if (BrowserInformationControl.isAvailable(parent)) {
+                val tbm = ToolBarManager(8388608)
+                val font = "org.eclipse.jdt.ui.javadocfont"
+                BrowserInformationControl(parent, font, tbm)
+            } else {
+                DefaultInformationControl(parent, true)
+            }
+    }
+
+    private val fHoverControlCreator: JavadocHover.HoverControlCreator by lazy {
+        JavadocHover.HoverControlCreator(KotlinInformationControlCreator())
+    }
 
     override val hoverPriority: Int
         get() = 2
 
-    companion object {
-        private val DUMMY_REGION = Region(0, 0)
-    }
-    
+    override fun getHoverControlCreator(): IInformationControlCreator = fHoverControlCreator
+
     override fun getHoverInfo(hoverData: HoverData): BrowserInformationControlInput? {
-        val (element, editor) = hoverData
-        
-        val javaElements = obtainJavaElements(element)
-        if (javaElements.isEmpty()) return null
-        
-        val editorInputElement = EditorUtility.getEditorInputJavaElement(editor.javaEditor, false)
-        
-        return JavadocHover.getHoverInfo(javaElements.toTypedArray(), editorInputElement, DUMMY_REGION, null)
+        val (element, _) = hoverData
+        return KotlinAnnotationBrowserInformationControlInput(element)
```

Comment:
No reason to use additional `val`: `KotlinAnnotationBrowserInformationControlInput(hoverData.hoverElement)`